sal-service/src/main/java/org/ow2/proactive/sal/service/service/CloudService.java

```diff
@@ -239,17 +243,25 @@ public Boolean removeClouds(String sessionId, List<String> cloudIds, Boolean pre
                              cloud.getCloudType());
             }
             if (Boolean.TRUE.equals(tempFlag)) {
+                List<Deployment> toBeRemovedDeployments = new ArrayList<>();
                 if (cloud.getDeployments() != null) {
-                    LOGGER.info("Cleaning deployments from related tasks " + cloud.getDeployments().toString());
-                    cloud.getDeployments().forEach(deployment -> deployment.getTask().removeDeployment(deployment));
+                    LOGGER.info("Cleaning all deployments related to cloud \"{}\"", cloud.getCloudId());
+                    toBeRemovedDeployments.addAll(cloud.getDeployments());
+                    for (Deployment deployment : toBeRemovedDeployments) {
+                        repositoryService.deleteDeployment(deployment);
+                    }
```

Comment:
No need to use a new list here + check if it empty so we don't enter block if yes.

```
if (cloud.getDeployments() != null && !cloud.getDeployments().isEmpty()) {
    LOGGER.info("Cleaning all deployments related to cloud \"{}\"", cloud.getCloudId());
    for (Deployment deployment : cloud.getDeployments()) {
        repositoryService.deleteDeployment(deployment);
     }
}
```

Reply:
I agree with checking whether the list is empty.
However, yes you need to create a copy of the list since the elements of the list are being deleted. A list variable is a pointer to the first element of the list, in this case, the moment I delete the first element,  the variable will be pointing to null. So, yes there is a need to create a copy ;)

Reply:
Ah its removed from the list in `repositoryService.deleteDeployment(deployment);`.
My bad you are right you need a copy.